toolchain/check/facet_type.cpp

```diff
@@ -281,6 +283,126 @@ auto IsPeriodSelf(Context& context, SemIR::InstId inst_id) -> bool {
+  auto lhs_access = context.insts().TryGetAs<SemIR::ImplWitnessAccess>(inst_id);
+  if (!lhs_access) {
+    return std::nullopt;
+  }
+  auto lhs_lookup = context.insts().TryGetAs<SemIR::LookupImplWitness>(
+      lhs_access->witness_id);
+  if (!lhs_lookup) {
+    return std::nullopt;
+  }
+  if (!IsPeriodSelf(context, lhs_lookup->query_self_inst_id)) {
+    return std::nullopt;
+  }
+  return lhs_access;
+}
+
+// To be used for substituting into the RHS of a rewrite constraint.
+//
+// It will substitute any `ImplWitnessAccess` into `.Self` (a reference to an
+// associated constant) with the RHS of another rewrite constraint that writes
+// to the same associated constant. For example:
+// ```
+// Z where .X = () and .Y = .X
+// ```
+// Here the second `.X` is an `ImplWitnessAccess` which would be substituted by
+// finding the first rewrite constraint, where the LHS is for the same
+// associated constant and using its RHS. So the substitution would produce:
+// ```
+// Z where .X = () and .Y = ()
+// ```
+//
+// This additionally diagnoses cycles when the `ImplWitnessAccess` is reading
+// from the same rewrite constraint, and is thus assigning to the associated
+// constant a value that refers to the same associated constant, such as with
+// `Z where .X = C(.X)`. In the event of a cycle, the `ImplWitnessAccess` is
+// replaced with `ErrorInst` so that further evaluation of the
+// `ImplWitnessAccess` will not loop infinitely.
+class SubstImplWitnessAccessCallbacks : public SubstInstCallbacks {
+ public:
+  explicit SubstImplWitnessAccessCallbacks(
+      Context* context, SemIR::LocId loc_id,
+      const SemIR::FacetTypeInfo* facet_type,
+      const SemIR::FacetTypeInfo::RewriteConstraint* current_constraint)
+      : SubstInstCallbacks(context),
+        loc_id_(loc_id),
+        facet_type_(facet_type),
+        current_constraint_(current_constraint) {}
+
+  auto Subst(SemIR::InstId& rhs_inst_id) const -> bool override {
+    if (context().constant_values().Get(rhs_inst_id).is_concrete()) {
+      return true;
+    }
+
+    auto rhs_access =
+        TryGetImplWitnessAccessOfPeriodSelf(context(), rhs_inst_id);
+    if (!rhs_access) {
+      return false;
+    }
+
+    // TODO: We could consider replacing this loop, which makes for an O(N^2)
+    // algorithm with something more efficient for searching, such as a map.
+    // However that would probably require heap allocations which may be worse
+    // overall since the number of rewrite constraints is generally low.
+    for (const auto& search_constraint : facet_type().rewrite_constraints) {
+      auto search_lhs_access = TryGetImplWitnessAccessOfPeriodSelf(
+          context(), search_constraint.lhs_id);
+      if (!search_lhs_access) {
+        continue;
+      }
+
+      if (search_lhs_access->witness_id == rhs_access->witness_id &&
+          search_lhs_access->index == rhs_access->index) {
+        if (IsCurrentConstraint(search_constraint)) {
```

Comment:
I think this would actually be a little clearer without the indirection of a function call.

```suggestion
        if (&search_constraint == current_constraint_) {
```

Reply:
Done